### ISSUE TITLE
Unexpected 'never' errors when using mapped types

### ISSUE THREAD

**Issue by OliverJAsh** [Microsoft/TypeScript#20035]
**TypeScript Version:**  2.6.1

**Code**

```ts
{
  const getKey = function<Key extends string, X> (
    key: Key,
    xs: { [key in Key]: X }
  ): X | undefined {
    return xs[key];
  };

  const record = {} as Record<string, string>;
  const record2 = {} as { [key: string]: string };

  // Unexpected error: Type 'string' is not assignable to type 'never'
  getKey('foo', record);
  // Unexpected error: Type 'string' is not assignable to type 'never'
  getKey('foo', record2);
}
```

If I change the type of the `xs` parameter to `Record<Key, X>`, this seems to help, although not completely:

``` ts
{
  const getKey = function<Key extends string, X> (
    key: Key,
    xs: Record<Key, X>
  ): X | undefined {
    return xs[key];
  };

  const record = {} as Record<string, string>;
  const record2 = {} as { [key: string]: string };

  // No error, good!
  getKey('foo', record);
  // Unexpected error: Type 'string' is not assignable to type 'never'
  getKey('foo', record2);
}
```

I suspect this compiler error is a false positive?

I'm also not sure how using `Record` helps, as the type definition appears identical to me.

**Comment by OliverJAsh** [comment]
My mistake. I was confusing records and maps.

Instead of using mapped types, I needed index types:

``` ts
const getKeyInMap = function<X> (
  key: string,
  xs: { [key: string]: X },
): X | undefined {
  return xs[key];
};
```
